execution/src/vm.rs

```diff
@@ -0,0 +1,298 @@
+                        (Some(_), None) => None, // block/miss mismatch
+                    }
+                }
+            } else {
+                None
+            }
+        };
+        if let Some(cached) = opt_cached {
+            // execution was already done, apply cached ledger changes to final ledger
+            let mut final_ledger_guard = self.final_ledger.lock().unwrap();
+            (*final_ledger_guard).apply_changes(&cached);
+            return;
+        }
+
+        // nothing found in cache, or cache mismatch: reset history, run step and make it final
+        // this should almost never happen, so the heavy step.clone() is OK
+        self.reset_to_final();
+        self.run_active_step(step.clone());
+        self.run_final_step(step);
+    }
+
+    /// runs an SCE-active execution step
+    /// # Parameters
+    ///   * step: execution step to run
+    pub fn run_active_step(&mut self, step: ExecutionStep) {
+        // accumulate active ledger changes history
+        let mut active_changes = SCELedgerChanges::default();
+        for (_step_slot, _step_opt_block, step_changes) in self.step_history.iter() {
+            active_changes.apply_changes(step_changes);
+        }
+
+        // setup active execution context
+        {
+            let mut exec_context_guard = self.current_execution_context.0.lock().unwrap();
+            // clear caused changes
+            (*exec_context_guard).ledger_step.caused_changes.0.clear();
+            // accumulated step history
+            (*exec_context_guard).ledger_step.cumulative_history_changes = active_changes;
+            // TODO add more info in the exec_context: slot, PoS draws, optional block, call stack etc...
+            //      if possible prefer Arc/Mutex sharing to copying full blocks
+        }
+
+        // run implicit and async calls
+        //TODO
+
+        // run explicit calls within the block (if the slot is not a miss)
+        // note that total block gas is not checked, because currently Protocol makes the block invalid if it overflows gas
+        let opt_block_id: Option<BlockId>;
+        if let Some((block_id, block)) = step.block {
+            opt_block_id = Some(block_id);
+
+            // get block creator addr
+            // TODO remove unwrap
+            let block_creator_addr =
+                Address::from_public_key(&block.header.content.creator).unwrap();
+
+            // run all operations
+            for (op_idx, operation) in block.operations.into_iter().enumerate() {
+                let (module, max_gas, coins, gas_price, init_changes, sender_addr) =
+                    if let OperationType::ExecuteSC {
+                        data,
+                        max_gas,
+                        coins,
+                        gas_price,
+                    } = operation.content.op
+                    {
+                        // get sender address
+                        // TODO remove unwrap
```

Comment:
Maybe we can try to use this error as promoted by @yvan-sraka

Reply:
yes it's an option, for now I haven't written a good error handling here yet